lxd/db/storage_pools.go

```diff
@@ -107,13 +114,35 @@ SELECT storage_volumes.name, storage_volumes.type, projects.name, storage_volume
 		return nil, err
 	}
 
+	remoteVols := make(map[string]struct{})
 	for _, r := range vols {
+
+		// Deduplicate remote volumes.
+		if allNodes {
+			if r.nodeID == nil {
+				_, ok := remoteVols[r.volName]
+				if !ok {
+					remoteVols[r.volName] = struct{}{}
+				} else {
+					continue
+				}
+			}
+		}
+
```

Comment:
@stgraber @tomponline as far as I can tell we don't need to deduplicate the remote volumes at all:

The `storage_volumes` table allows the `node_id` column to be null. Remote volumes are given a null `node_id` and thus are not duplicated in the table itself:
```
root@node1:~# lxd sql global 'select id, name, node_id from storage_volumes'
+----+------------------------------------------------------------------+---------+
| id |                               name                               | node_id |
+----+------------------------------------------------------------------+---------+
| 1  | c1                                                               | 1       |
| 3  | 859a0a6696a0ef4554edb9d4519dbd73fcf268512472a353afaca5c6fbe96eee | <nil>   |
| 2  | c2                                                               | <nil>   |
| 4  | c3                                                               | <nil>   |
+----+------------------------------------------------------------------+---------+
```

This means that the above implementation for getting the used-by data works just fine without deduplication on the volumes. In fact, since volume names are not enforced to be unique in the current schema, deduplicating on volume name may have led to errors.

I think there are a few options here:
1. Assume that the `node_id` of the volumes will always be null for remote storage. If we make this assumption we can simplify the query somewhat. I think this will be fine, there have been schema updates to enforce it (https://github.com/lxc/lxd/blob/master/lxd/db/cluster/update.go#L1650 & https://github.com/lxc/lxd/blob/master/lxd/db/cluster/update.go#L1622).
2. Make no assumptions and deduplicate the volumes based on their id.
3. Same as 2. but ensure that the driver is not a remote driver if the node id is not null when querying the db.

Reply:
Just to clarify, with option 1 the where clause of the query would be simplified to:
```go
nodePredicate := ""
if !allNodes {
    nodePredicate = "AND node_id=?"
}
whereClause := fmt.Sprintf("WHERE storage_pool_id=? %s OR storage_volumes.type == 2", nodePredicate)
```

Reply:
I'll recheck on a long lived cluster but it very well may be that we fixed that database mess some time ago and are indeed using a NULL node_id everywhere now to avoid the duplicated rows.

Reply:
Confirmed that on a busy cluster that's ceph backed, all volumes have NULL as their node_id.

Reply:
Ok cool, I've removed the deduplication and simplified the query.